genesis/parsing/smartContractParser.go

```diff
@@ -0,0 +1,166 @@
+	"github.com/ElrondNetwork/elrond-go/genesis/data"
+	"github.com/ElrondNetwork/elrond-go/sharding"
+)
+
+// smartContractParser hold data for initial smart contracts
+type smartContractParser struct {
+	initialSmartContracts []*data.InitialSmartContract
+	pubkeyConverter       state.PubkeyConverter
+	checkForFileHandler   func(filename string) error
+}
+
+// NewSmartContractsParser creates a new decoded smart contracts genesis structure from json config file
+func NewSmartContractsParser(
+	genesisFilePath string,
+	pubkeyConverter state.PubkeyConverter,
+) (*smartContractParser, error) {
+
+	if check.IfNil(pubkeyConverter) {
+		return nil, genesis.ErrNilPubkeyConverter
+	}
+
+	initialSmartContracts := make([]*data.InitialSmartContract, 0)
+	err := core.LoadJsonFile(&initialSmartContracts, genesisFilePath)
+	if err != nil {
+		return nil, err
+	}
+
+	scp := &smartContractParser{
+		initialSmartContracts: initialSmartContracts,
+		pubkeyConverter:       pubkeyConverter,
+	}
+	scp.checkForFileHandler = scp.checkForFile
+
+	err = scp.process()
+	if err != nil {
+		return nil, err
+	}
+
+	return scp, nil
+}
+
+func (scp *smartContractParser) process() error {
+	for _, initialSmartContract := range scp.initialSmartContracts {
+		err := scp.parseElement(initialSmartContract)
+		if err != nil {
+			return err
+		}
+
+		err = scp.checkForFileHandler(initialSmartContract.Filename)
+		if err != nil {
+			return err
+		}
+	}
+
+	err := scp.checkForDuplicates()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (scp *smartContractParser) parseElement(initialSmartContract *data.InitialSmartContract) error {
+	if len(initialSmartContract.Owner) == 0 {
+		return genesis.ErrEmptyOwnerAddress
+	}
+	ownerBytes, err := scp.pubkeyConverter.Decode(initialSmartContract.Owner)
+	if err != nil {
+		return fmt.Errorf("%w for `%s`",
+			genesis.ErrInvalidOwnerAddress, initialSmartContract.Owner)
+	}
+
+	initialSmartContract.SetOwnerBytes(ownerBytes)
+
+	if len(initialSmartContract.VmType) == 0 {
+		return fmt.Errorf("%w for  %s",
+			genesis.ErrEmptyVmType, initialSmartContract.Owner)
+	}
+
+	_, err = hex.DecodeString(initialSmartContract.VmType)
+	if err != nil {
+		return fmt.Errorf("%w for provided %s, error: %s",
+			genesis.ErrInvalidVmType, initialSmartContract.VmType, err.Error())
+	}
+
+	return nil
+}
+
+func (scp *smartContractParser) checkForFile(filename string) error {
+	info, err := os.Stat(filename)
+	if os.IsNotExist(err) {
+		return fmt.Errorf("%w for the file %s", err, filename)
+	}
+
+	if info.IsDir() {
+		return fmt.Errorf("%w for the file %s", genesis.ErrFilenameIsDirectory, filename)
+	}
+
+	return nil
+}
+
+func (scp *smartContractParser) checkForDuplicates() error {
+	for idx1 := 0; idx1 < len(scp.initialSmartContracts); idx1++ {
+		ib1 := scp.initialSmartContracts[idx1]
+		for idx2 := idx1 + 1; idx2 < len(scp.initialSmartContracts); idx2++ {
+			ib2 := scp.initialSmartContracts[idx2]
+			if ib1.Owner == ib2.Owner {
```

Comment:
what if elrond wants to deploy a set of contracts ? it should be not a problem - as you can do it with different nonces.

Reply:
Agree, removed the functions, allowed to do multiple deployments